OSDOCS 13097:Added a new section about override scheduling feature for cert-manager (#87192)
openshift/openshift-docs: modules/cert-manager-override-scheduling.adoc

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-customizing-api-fields.adoc

:_mod-docs-content-type: PROCEDURE
[id="cert-manager-override-scheduling_{context}"]
= Configuring scheduling overrides for cert-manager components

You can configure the pod scheduling from the {cert-manager-operator} API for the {cert-manager-operator} components such as cert-manager controller, CA injector, and Webhook. 

.Prerequisites

* You have access to the {product-title} cluster as a user with the `cluster-admin` role.
* You have installed version 1.15.0 or later of the {cert-manager-operator}.

.Procedure

> * Update the `certmanager.operator` custom resource to configure pod scheduling overrides for the desired components. Use the `overrideScheduling` field under the `controllerConfig`, `webhookConfig`, or `cainjectorConfig` sections to define `nodeSelector` and `tolerations` settings.

Reviewer: Still just need to add a statement about running a command somewhere in here:
```suggestion
* Update the `certmanager.operator` custom resource to configure pod scheduling overrides for the desired components by running the following command. Use the `overrideScheduling` field under the `controllerConfig`, `webhookConfig`, or `cainjectorConfig` sections to define `nodeSelector` and `tolerations` settings.
```

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-customizing-api-fields.adoc

:_mod-docs-content-type: PROCEDURE
[id="cert-manager-override-scheduling_{context}"]
= Configuring scheduling overrides for cert-manager components

> After installing the {cert-manager-operator}, you can configure the pod scheduling from the {cert-manager-operator} API for the {cert-manager-operator} components such as cert-manager controller, CA injector, and Webhook. 

Author: Agreed. The first option works well.

.Prerequisites

* You have access to the {product-title} cluster as a user with the `cluster-admin` role.
* You have installed version 1.15.0 or later of the {cert-manager-operator}.

.Procedure

* Update the `certmanager.operator` custom resource to configure pod scheduling overrides for the desired components. Use the `overrideScheduling` field under `controllerConfig`, `webhookConfig`, or `cainjectorConfig` to define `nodeSelector` and `tolerations` settings.
+
.Example output
[source,yaml]
----

> $ oc patch certmanager.operator cluster --type=merge -p="

Reviewer 2: yes, we definitely need to!

        node-role.kubernetes.io/control-plane: '' <1>
      tolerations:
        - key: node-role.kubernetes.io/master
          operator: Exists
          effect: NoSchedule <2>
  webhookConfig:
    overrideScheduling:
      nodeSelector:
        node-role.kubernetes.io/control-plane: '' <3>
      tolerations:
        - key: node-role.kubernetes.io/master
          operator: Exists
          effect: NoSchedule <4>
  cainjectorConfig:
    overrideScheduling:
      nodeSelector:
        node-role.kubernetes.io/control-plane: '' <5>
      tolerations:
        - key: node-role.kubernetes.io/master
          operator: Exists
          effect: NoSchedule <6>
----
<1> Defines the `nodeSelector` for the cert-manager controller deployment.
<2> Defines the `toleration` for the cert-manager controller deployment.
<3> Defines the `nodeSelector` for the cert-manager webhook deployment.
<4> Defines the `toleration` for the cert-manager webhook deployment.
<5> Defines the `nodeSelector` for the cert-manager cainjector deployment.
<6> Defines the `toleration` for the cert-manager cainjector deployment.

.Verification

. Verify pod scheduling settings for `cert-manager` pods:

> .. Check the deployments in the `cert-manager` namespace to confirm they have the correct `nodeSelector` and `tolerations`:

Reviewer: As mentioned before, steps that introduce commands need to specifically mention that:
```suggestion
.. Check the deployments in the `cert-manager` namespace to confirm they have the correct `nodeSelector` and `tolerations` by running the following command:
```

+
[source,terminal]
----
$ oc get pods -n cert-manager -o wide
----
+
.Example output
[source,yaml]
----
NAME                                       READY   STATUS    RESTARTS   AGE   IP            NODE                         NOMINATED NODE   READINESS GATES
cert-manager-58d9c69db4-78mzp              1/1     Running   0          10m   10.129.0.36   ip-10-0-1-106.ec2.internal   <none>           <none>
cert-manager-cainjector-85b6987c66-rhzf7   1/1     Running   0          11m   10.128.0.39   ip-10-0-1-136.ec2.internal   <none>           <none>
cert-manager-webhook-7f54b4b858-29bsp      1/1     Running   0          11m   10.129.0.35   ip-10-0-1-106.ec2.internal   <none>           <none>
----

> .. To check the `nodeSelector` and `tolerations` settings applied to deployments, run the following:

Reviewer: Per the [ISG](https://www.ibm.com/docs/en/ibm-style?topic=format-procedures#writing-steps), "Use imperative mood in the first sentence of each step."
```suggestion
.. Check the `nodeSelector` and `tolerations` settings applied to deployments by running the following command:
```

+
[source,terminal]
----
$ oc get deployments -n cert-manager -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}{.spec.template.spec.nodeSelector}{"\n"}{.spec.template.spec.tolerations}{"\n\n"}{end}'
----
+

> .Exaple output

Reviewer: ```suggestion
.Example output
```

> [source,yaml]

Reviewer: Same comment about this being terminal output:
```suggestion
[source,terminal]
```

----
cert-manager
{"kubernetes.io/os":"linux","node-role.kubernetes.io/control-plane":""}
[{"effect":"NoSchedule","key":"node-role.kubernetes.io/master","operator":"Exists"}]

cert-manager-cainjector
{"kubernetes.io/os":"linux","node-role.kubernetes.io/control-plane":""}
[{"effect":"NoSchedule","key":"node-role.kubernetes.io/master","operator":"Exists"}]

cert-manager-webhook
{"kubernetes.io/os":"linux","node-role.kubernetes.io/control-plane":""}
[{"effect":"NoSchedule","key":"node-role.kubernetes.io/master","operator":"Exists"}]
----

> . Verify pod scheduling events in the `cert-manager` namespace:

Reviewer: ```suggestion
. Verify pod scheduling events in the `cert-manager` namespace by running the following command:
```

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-customizing-api-fields.adoc

:_mod-docs-content-type: PROCEDURE
[id="cert-manager-override-scheduling_{context}"]

> = Configuring Scheduling Overrides for {cert-manager} Components

Reviewer 2: the `{cert-manager}` placeholder doesn't get rendered into a value!

> After installing the {cert-manager-operator}, you can configure the scheduling of {cert-manager} components by configuring the `overrideScheduling` field in the {cert-manager} custom resource. This allows you to define `nodeSelector` and `tolerations` for the `controllerConfig`, `webhookConfig`, and `cainjectorConfig` sections.

Reviewer 3: ```suggestion
After installing the {cert-manager-operator}, you can configure the scheduling of cert-manager components by configuring the `overrideScheduling` field in the `CertManager` custom resource. This allows you to deploy the operands such as cert-manager controller, CA injector, and webhook pods onto specific nodes with matched `nodeSelector` and `tolerations` configurations.
```

Author: Ack.

> After installing the {cert-manager-operator}, you can configure the scheduling of {cert-manager} components by configuring the `overrideScheduling` field in the {cert-manager} custom resource. This allows you to define `nodeSelector` and `tolerations` for the `controllerConfig`, `webhookConfig`, and `cainjectorConfig` sections.

Reviewer 2: same here, while `{cert-manager-operator}` does the other `{cert-manager}` stays as is in the doc page.

Reviewer 2: ```suggestion
After installing the {cert-manager-operator}, you can configure the pod scheduling from the {cert-manager-operator} API for the cert-manager components such as cert-manager controller, CA injector, and Webhook. Configuring the `overrideScheduling` block in the certmanager.operator custom resource allows you to define `nodeSelector` and `tolerations` for the `controllerConfig`, `webhookConfig`, and `cainjectorConfig` sections respectively.
```

.Prerequisites

* You have access to the {product-title} cluster as a user with the `cluster-admin` role.

> * You have installed the {cert-manager-operator} 1.12.0 or later.

Reviewer 3: ```suggestion
* You have installed the {cert-manager-operator} 1.15.0 or later.
```

.Procedure

> * To configure scheduling overrides, modify the {cert-manager} custom resource. Use the `overrideScheduling` field within `controllerConfig`, `webhookConfig`, or `cainjectorConfig` to specify `nodeSelector` and `tolerations` for the respective component.

Reviewer 3: ```suggestion
* To configure scheduling overrides, modify the `CertManager` custom resource. Use the `overrideScheduling` field within `controllerConfig`, `webhookConfig`, or `cainjectorConfig` to specify `nodeSelector` and `tolerations` for the respective component.
```

Reviewer 2: ++

> * To configure scheduling overrides, modify the {cert-manager} custom resource. Use the `overrideScheduling` field within `controllerConfig`, `webhookConfig`, or `cainjectorConfig` to specify `nodeSelector` and `tolerations` for the respective component.

Reviewer 2: another instance of `{cert-manager}

> * To configure scheduling overrides, modify the {cert-manager} custom resource. Use the `overrideScheduling` field within `controllerConfig`, `webhookConfig`, or `cainjectorConfig` to specify `nodeSelector` and `tolerations` for the respective component.

Reviewer 2: ```suggestion
* To configure scheduling overrides, modify the certmanager.operator custom resource. Use the `overrideScheduling` field within `controllerConfig`, `webhookConfig`, or `cainjectorConfig` to specify `nodeSelector` and `tolerations` for overriding scheduling of cert-manager controller, webhook or cainjector pods respectively.
```

+
.Example output
[source,yaml]
----
apiVersion: operator.openshift.io/v1alpha1
kind: CertManager
metadata:
  name: cluster
spec:
  controllerConfig:
    overrideScheduling:
      nodeSelector:
        <key1>: <value1> <1>
        <key2>: <value2>
      tolerations:
        - key: "<key1>" <1>
          operator: "Equal"
          value: "<value1>"
          effect: "NoSchedule"
  webhookConfig:
    overrideScheduling:
      nodeSelector:
        <key1>: <value1>
      tolerations:
        - key: "<key1>"
          operator: "Equal"
          value: "<value1>"
          effect: "NoSchedule"
  cainjectorConfig:
    overrideScheduling:
      nodeSelector:
        <key1>: <value1>
      tolerations:
        - key: "<key1>"
          operator: "Equal"
          value: "<value1>"
          effect: "NoSchedule"
----

> <1> Replace <key1>, <key2> and <value1> with the specific key-value pairs for `nodeSelector` and `tolerations`.

Reviewer 2: do we need to add examples for these?

Author: If we already have examples, we can add it.

Reviewer 3: How about using the example documented in [OCP-77779](https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-77779), to schedule the pods onto master (control-plane) nodes.

```
$ oc patch certmanager.operator cluster --type=merge -p="
spec:
  controllerConfig:
    overrideScheduling:
      nodeSelector:
        node-role.kubernetes.io/control-plane: ''
      tolerations:
        - key: node-role.kubernetes.io/master
          operator: Exists
          effect: NoSchedule
  webhookConfig:
    overrideScheduling:
      nodeSelector:
        node-role.kubernetes.io/control-plane: ''
      tolerations:
        - key: node-role.kubernetes.io/master
          operator: Exists
          effect: NoSchedule
  cainjectorConfig:
    overrideScheduling:
      nodeSelector:
        node-role.kubernetes.io/control-plane: ''
      tolerations:
        - key: node-role.kubernetes.io/master
          operator: Exists
          effect: NoSchedule
"
```

.Verification

* After applying the updated {cert-manager} custom resource, verify that the scheduling overrides are active:

> 

Reviewer 2: we can add two verification steps,

1. checking the deployment(s) in the cert-manager namespace
    - if the deployment contains the right nodeSelector and tolerations
2. checking the pod(s) in the cert-manager namespace
    - if they are scheduled on the right nodes

Reviewer 2: `$ oc get deployments -n cert-manager -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}{.spec.template.spec.nodeSelector}{"\n"}{.spec.template.spec.tolerations}{"\n\n"}{end}'`

for Verification step 1.

> .. Check the pod specifications for the {cert-manager} components:

Reviewer 2: same, `{cert-manager}`

+
[source,terminal]
----

> $ kubectl get pods -n openshift-cert-manager -o yaml | grep -A 10 "nodeSelector"

Reviewer 3: ```suggestion
$ oc get deployment -n cert-manager -o yaml
```

Reviewer 2: ++

> $ kubectl get pods -n openshift-cert-manager -o yaml | grep -A 10 "nodeSelector"

Reviewer 2: we prefer `oc` over `kubectl`, everywhere!

Reviewer 2: ```suggestion
$ oc get pods -n cert-manager -o wide
```
^ also, add an output please!

After which, we can say that "verify if pods were scheduled on the desired nodes"...

Reviewer 2: ^ suffices for Verification step 2.

----

> .. Confirm that the `nodeSelector` and `tolerations` settings match the configuration.

Reviewer 2: can be part of Verification step 1.

// Module included in the following assemblies:
//
// * security/cert_manager_operator/cert-manager-customizing-api-fields.adoc

:_mod-docs-content-type: PROCEDURE
[id="cert-manager-override-scheduling_{context}"]
= Configuring scheduling overrides for cert-manager Operator components

> After installing the {cert-manager-operator}, you can configure the pod scheduling from the {cert-manager-operator} API for the {cert-manager-operator} components such as {cert-manager-operator} controller, CA injector, and Webhook. Configuring the `overrideScheduling` block in the `certmanager.operator` custom resource enables you to define `nodeSelector` and `tolerations` for the `controllerConfig`, `webhookConfig`, and `cainjectorConfig` sections respectively.
> 

Reviewer 3: To be more accurate and simple:
```suggestion
After installing the {cert-manager-operator}, you can configure the pod scheduling from the {cert-manager-operator} API for the cert-manager components such as cert-manager controller, CA injector, and Webhook. Configuring the `overrideScheduling` block in the `certmanager.operator` custom resource enables you to define `nodeSelector` and `tolerations` for the `controllerConfig`, `webhookConfig`, and `cainjectorConfig` sections respectively.
```

.Prerequisites

* You have access to the {product-title} cluster as a user with the `cluster-admin` role.

> * You have installed the {cert-manager-operator} 1.15.0 or later.

Reviewer 3: Just to keep consistent with existing words.
```suggestion
* You have installed version 1.15.0 or later of the {cert-manager-operator}.
```

.Procedure

> * To configure scheduling overrides, modify the `certmanager.operator` custom resource. Use the `overrideScheduling` field within `controllerConfig`, `webhookConfig`, or `cainjectorConfig` to specify `nodeSelector` and `tolerations` for overriding scheduling of {cert-manager-operator} `controller`, `webhook` or `cainjector` pods respectively.

Reviewer 3: There are some wording duplicates with Line 9: https://github.com/openshift/openshift-docs/pull/87192/files#diff-231d428c93d050aa8bbff9119037005068aa3c57299adbfc8c708db81ba9602dR9

> * To configure scheduling overrides, modify the `certmanager.operator` custom resource. Use the `overrideScheduling` field within `controllerConfig`, `webhookConfig`, or `cainjectorConfig` to specify `nodeSelector` and `tolerations` for overriding scheduling of {cert-manager-operator} `controller`, `webhook` or `cainjector` pods respectively.

Reviewer 3: There are some duplicates with Line 9: https://github.com/openshift/openshift-docs/pull/87192/files#diff-231d428c93d050aa8bbff9119037005068aa3c57299adbfc8c708db81ba9602dR9

+

> .Example output

Reviewer 3: It's a part of the `Procedure`

```suggestion
```

Reviewer: +1 this needs to be removed since this is not an output, as far as I can tell, this is the actual command that users need to run

[source,yaml]
----
$ oc patch certmanager.operator cluster --type=merge -p="
spec:
  controllerConfig:
    overrideScheduling:
      nodeSelector:
        node-role.kubernetes.io/control-plane: '' <1>
      tolerations:
        - key: node-role.kubernetes.io/master
          operator: Exists
          effect: NoSchedule <2>
  webhookConfig:
    overrideScheduling:
      nodeSelector:
        node-role.kubernetes.io/control-plane: '' <3>
      tolerations:
        - key: node-role.kubernetes.io/master
          operator: Exists
          effect: NoSchedule <4>
  cainjectorConfig:
    overrideScheduling:
      nodeSelector:
        node-role.kubernetes.io/control-plane: '' <5>
      tolerations:
        - key: node-role.kubernetes.io/master
          operator: Exists
          effect: NoSchedule <6>
----

> <1> Defines the control-plane `nodeSelection` for the `controllerConfig` override scheduling.

Reviewer 2: ```suggestion
<1> Defines the `nodeSelector` for the cert-manager controller deployment.
```

> <2> Defines the `toleration` for `controllerConfig`.

Reviewer 2: ```suggestion
<2> Defines the `toleration` for the cert-manager controller deployment.
```

> <3> Defines the `nodeSelector` for `webhookConfig`.

Reviewer 2: ```suggestion
<3> Defines the `nodeSelector` for the cert-manager webhook deployment.
```

> <4> Defines the `toleration` for `webhookConfig`.

Reviewer 2: ```suggestion
<4> Defines the `toleration` for the cert-manager webhook deployment.
```

> <5> Defines the `nodeSelector` for `cainjectorConfig`.

Reviewer 2: ```suggestion
<5> Defines the `nodeSelector` for the cert-manager cainjector deployment.
```

> <6> Defines the `toleration` for `cainjectorConfig`.

Reviewer 2: ```suggestion
<6> Defines the `toleration` for the cert-manager cainjector deployment.
```

.Verification

> . Verify the updated `certmanager.operator` custom resource by:

Reviewer 2: We're not actually checking the `certmanager.operator` resource in this verification stage,
please rephrase this statement.

`Verify the pod scheduling changes have been applied` ? or something similar generic would help.

> * Check the deployments in the {cert-manager-operator} namespace to confirm they have the correct `nodeSelector` and `tolerations`:

Reviewer 2: ```suggestion
* Check the deployments in the `cert-manager` namespace to confirm they have the correct `nodeSelector` and `tolerations`:
```

+
[source,terminal]
----
$ oc get pods -n cert-manager -o wide
----
+
.Example output
[source,yaml]
----
$ oc get deployments -n cert-manager -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}{.spec.template.spec.nodeSelector}{"\n"}{.spec.template.spec.tolerations}{"\n\n"}{end}'

> ----

Reviewer 2: These should be two distinct steps:

1. command input: `$ oc get pods -n cert-manager -o wide`

example output:
```
NAME                                       READY   STATUS    RESTARTS   AGE   IP            NODE                         NOMINATED NODE   READINESS GATES
cert-manager-58d9c69db4-78mzp              1/1     Running   0          10m   10.129.0.36   ip-10-0-1-106.ec2.internal   <none>           <none>
cert-manager-cainjector-85b6987c66-rhzf7   1/1     Running   0          11m   10.128.0.39   ip-10-0-1-136.ec2.internal   <none>           <none>
cert-manager-webhook-7f54b4b858-29bsp      1/1     Running   0          11m   10.129.0.35   ip-10-0-1-106.ec2.internal   <none>           <none>
```

2. command input: `$ oc get deployments -n cert-manager -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}{.spec.template.spec.nodeSelector}{"\n"}{.spec.template.spec.tolerations}{"\n\n"}{end}'`

example output:
```
cert-manager
{"kubernetes.io/os":"linux","node-role.kubernetes.io/control-plane":""}
[{"effect":"NoSchedule","key":"node-role.kubernetes.io/master","operator":"Exists"}]

cert-manager-cainjector
{"kubernetes.io/os":"linux","node-role.kubernetes.io/control-plane":""}
[{"effect":"NoSchedule","key":"node-role.kubernetes.io/master","operator":"Exists"}]

cert-manager-webhook
{"kubernetes.io/os":"linux","node-role.kubernetes.io/control-plane":""}
[{"effect":"NoSchedule","key":"node-role.kubernetes.io/master","operator":"Exists"}]
```

3. `$ oc get events -n cert-manager --field-selector reason=Scheduled` (command for that step seems fine ✅ )

> . Verify pod scheduling in the {cert-manager-operator} namespace:

Reviewer 2: ```suggestion
. Verify pod scheduling events in the `cert-manager` namespace:
```
